borsh-ts/serialize.ts

```diff
@@ -0,0 +1,176 @@
+import { ArrayType, MapType, IntegerType, OptionType, Schema, SetType, StructType, integers, EnumType } from './types.js';
+import { EncodeBuffer } from './buffer.js';
+import * as utils from './utils.js';
+
+export class BorshSerializer {
+    encoded: EncodeBuffer;
+    fieldPath: string[];
+
+    constructor() {
+        this.encoded = new EncodeBuffer();
+        this.fieldPath = ['value'];
+    }
+
+    encode(value: unknown, schema: Schema): Uint8Array {
+        this.encode_value(value, schema);
+        return this.encoded.get_used_buffer();
+    }
+
+    encode_value(value: unknown, schema: Schema): void {
+        if (typeof schema === 'string') {
+            if (integers.includes(schema)) return this.encode_integer(value, schema);
+            if (schema === 'string') return this.encode_string(value);
+            if (schema === 'bool') return this.encode_boolean(value);
+        }
+
+        if (typeof schema === 'object') {
+            if ('option' in schema) return this.encode_option(value, schema as OptionType);
+            if ('enum' in schema) return this.encode_enum(value, schema as EnumType);
+            if ('array' in schema) return this.encode_array(value, schema as ArrayType);
+            if ('set' in schema) return this.encode_set(value, schema as SetType);
+            if ('map' in schema) return this.encode_map(value, schema as MapType);
+            if ('struct' in schema) return this.encode_struct(value, schema as StructType);
+        }
+    }
+
+    encode_integer(value: unknown, schema: IntegerType): void {
+        const size: number = parseInt(schema.substring(1));
+
+        if (size <= 32 || schema == 'f64') {
+            utils.expect_type(value, 'number', this.fieldPath);
+            this.encoded.store_value(value as number, schema);
+        } else {
+            utils.expect_BN(value, this.fieldPath);
+            value = value instanceof BigInt ? value : BigInt(value as number | string);
```

Comment:
This may be replace with `BigInt(value.toString())`. Although BN it's not imported here, but we can keep a little backward-compatibility by using, in case the user pass a BN object:
<img width="128" alt="image" src="https://github.com/near/borsh-js/assets/13259400/d6a26080-003a-4514-b1b4-648509c94ee5">

Reply:
done, turns out we only need to do `BigInt(value)` and it will automatically try to call `toString` on the `value`.

Reply:
Nice! That's neat!